darwin/CpuFreq.c

```diff
@@ -0,0 +1,176 @@
+#include "darwin/CpuFreq.h"
+
+#ifdef CPUFREQ_SUPPORT
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+
+/* Implementations */
+int CpuFreq_init(const Machine* machine, CpuFreqData* data) {
+    data->existingCPUs = machine->existingCPUs;
+    data->cluster_type_per_cpu = xCalloc(data->existingCPUs, sizeof(uint32_t));
+    data->frequencies = xCalloc(data->existingCPUs, sizeof(double));
+
+    /* Determine the cluster type for all CPUs */
+    char buf[128];
+
+    for (uint32_t num_cpus = 0U; num_cpus < data->existingCPUs; num_cpus++) {
+        snprintf(buf, sizeof(buf), "IODeviceTree:/cpus/cpu%u", num_cpus);
+        io_registry_entry_t cpu_io_registry_entry = IORegistryEntryFromPath(kIOMainPortDefault, buf);
+        if (cpu_io_registry_entry == MACH_PORT_NULL) {
+            return -1;
```

Comment:
For this and all similar subsequent comments: the intention was that even on failure, we should be in a consistent state, so CpuFreq_cleanup can (and should) be called (and it is, if you look at DarwinMachine.c). If that's not preferred, I can refactor, but it will introduce more boilerplate error handling.

Reply:
Error handling is one of the rare cases where `goto err;` is actually a quite common pattern.